lib/flight_scheduler/schedulers/fifo_scheduler.rb

```diff
@@ -124,48 +44,89 @@ def allocate_jobs
     # jobs or find one that cannot be allocated.
 
     new_allocations = []
+    iterations = 0
+    max_iterations = 100
     @allocation_mutex.synchronize do
       loop do
-        # Select the next available job
-        next_group_id = @group_id_queue.detect do |id|
-          @data[id][:enum].peek
+        if iterations > max_iterations
+          break
+        else
+          iterations += 1
+        end
+
+        candidate = queue.detect do |job|
+          job.pending? && job.allocation.nil?
         end
-        break unless next_group_id
 
-        # Fetches the next job
-        # NOTE: Intentionally peek instead of next! The enumerator must not
-        #       progress until the allocation has been confirmed
-        enum = @data[next_group_id][:enum]
-        next_job = enum.peek
+        break if candidate.nil?
 
-        # Fast-Forward past allocated jobs
-        # TODO: Confirm if the is a valid use case or a by-product of the spec
-        #       Notionally jobs which are already allocated are being managed
-        #       by some external scheduler. This implies they shouldn't have
-        #       been added to this scheduler in the first place
-        if next_job.allocation
-          enum.next
-          next
+        if candidate.job_type == 'ARRAY_JOB'
```

Comment:
What does `next false` do?  Is that a typo or some Ruby that I've never seen before?

Reply:
The implementation of `queue` already has that check in it, but we are kind of duplicating it anyway....*thinks*.

Reply:
The `next` keyword allows you to specify a value for the current iteration of the block. It's just a habit when I've development when the return value of the block is important.

Reply:
So `next false unless ...` would cause the remainder of the block to be skipped and evaluate to `false` and for the next iteration to run.

Whereas `next unless ...` would cause the remainder of the block to be skipped and evaluate to `nil` and for the next iteration to run.

I suppose that could be useful in some cases.  Thanks for the answer.